internal/server/auth/method/github/server.go

```diff
@@ -0,0 +1,166 @@
+	config config.AuthenticationConfig,
+) *Server {
+	return &Server{
+		logger: logger,
+		store:  store,
+		config: config,
+		oauth2Config: &oauth2.Config{
+			ClientID:     config.Methods.Github.Method.ClientId,
+			ClientSecret: config.Methods.Github.Method.ClientSecret,
+			Endpoint:     oauth2GitHub.Endpoint,
+			RedirectURL:  callbackURL(config.Methods.Github.Method.RedirectAddress),
+			Scopes:       []string{strings.Join(config.Methods.Github.Method.Scopes, ":")},
+		},
+	}
+}
+
+// RegisterGRPC registers the server as an Server on the provided grpc server.
+func (s *Server) RegisterGRPC(server *grpc.Server) {
+	auth.RegisterAuthenticationMethodGithubServiceServer(server, s)
+}
+
+func callbackURL(host string) string {
+	// strip trailing slash from host
+	host = strings.TrimSuffix(host, "/")
+	return host + "/auth/v1/method/github/callback"
+}
+
+// AuthorizeURL will return a URL for the client to redirect to for completion of the OAuth flow with GitHub.
+func (s *Server) AuthorizeURL(ctx context.Context, req *auth.AuthorizeURLRequest) (*auth.AuthorizeURLResponse, error) {
+	u := s.oauth2Config.AuthCodeURL(req.State)
+
+	return &auth.AuthorizeURLResponse{
+		AuthorizeUrl: u,
+	}, nil
+}
+
+// Callback is the OAuth callback method for Github authentication. It will take in a Code
+// which is the OAuth grant passed in by the OAuth service, and exchange the grant with an Authentication
+// that includes the user information.
+func (s *Server) Callback(ctx context.Context, r *auth.CallbackRequest) (*auth.CallbackResponse, error) {
+	if r.State != "" {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return nil, errors.ErrUnauthenticatedf("missing state parameter")
+		}
+
+		state, ok := md["flipt_client_state"]
+		if !ok || len(state) < 1 {
+			return nil, errors.ErrUnauthenticatedf("missing state parameter")
+		}
+
+		if r.State != state[0] {
+			return nil, errors.ErrUnauthenticatedf("unexpected state parameter")
+		}
+	}
+
+	token, err := s.oauth2Config.Exchange(ctx, r.Code)
+	if err != nil {
+		return nil, err
+	}
+
+	if !token.Valid() {
+		return nil, errors.New("invalid token")
+	}
+
+	c := &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	userReq, err := http.NewRequestWithContext(ctx, "GET", githubUserAPI, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	userReq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.AccessToken))
+	userReq.Header.Set("Accept", "application/vnd.github+json")
+
+	var githubUserResponse struct {
+		Name  string `json:"name,omitempty"`
+		Email string `json:"email,omitempty"`
+	}
+
+	userResp, err := c.Do(userReq)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		userResp.Body.Close()
+	}()
+
+	if err := json.NewDecoder(userResp.Body).Decode(&githubUserResponse); err != nil {
+		return nil, err
+	}
+
+	metadata := map[string]string{
+		storageMetadataGithubEmail: githubUserResponse.Email,
+		storageMetadataGithubName:  githubUserResponse.Name,
+	}
```

Comment:
I am unsure how Github responds, but it might be nice to check if these are empty? (might need to confirm if these can ever be returned empty).
Then we should only set the keys if they're not empty.
Just incase the frontend something odd with empty values here.

Reply:
My one guess is email could be empty without the `user:email` scope.

Reply:
@GeorgeMac Something I was definitely considering too. I can do a quick test to see if that is the case, but I believe it might be.

Reply:
@GeorgeMac It seems as though this always gets populated without providing the `scope`. Kinda weird 🤔, i'll track down what the actual meaning of this is.

Reply:
I think it could be to do with your email visibility settings @yquansah 
If you go to Settings > Emails > Keep my email address private and toggle that in GH, it might change the output.
They might still stick your private github email address in the space, so that might not be a problem.